perfkitbenchmarker/providers/rackspace/rackspace_network.py

```
@@ -13,140 +13,295 @@
-                rule_cmd = []
-                rule_cmd.extend([FLAGS.neutron_path,
-                                 'security-group-rule-delete'])
-                rule_cmd.append(rule_id)
-
-                vm_util.IssueRetryableCommand(rule_cmd, env=firewall_env)
-
-            firewall_cmd = [FLAGS.neutron_path]
-            firewall_cmd.extend(['security-group-delete'])
-            firewall_cmd.append(firewall)
-
-            vm_util.IssueRetryableCommand(firewall_cmd, env=firewall_env)
-
-            self.firewall_names.remove(firewall)
+class RackspaceSecurityGroup(resource.BaseResource):
+  """An object representing a Rackspace Security Group."""
+
+  def __init__(self, name):
+    super(RackspaceSecurityGroup, self).__init__()
+    self.name = name
+    self.id = None
+
+  def _Create(self):
+    cmd = util.RackCLICommand(self, 'networks', 'security-group', 'create')
+    cmd.flags['name'] = self.name
+    stdout, stderr, _ = cmd.Issue()
+    resp = json.loads(stdout)
+    self.id = resp['ID']
+
+  def _Delete(self):
+    if self.id is None:
+      return
+    cmd = util.RackCLICommand(self, 'networks', 'security-group', 'delete')
+    cmd.flags['id'] = self.id
+    cmd.Issue()
+
+  def _Exists(self):
+    if self.id is None:
+      return False
+    cmd = util.RackCLICommand(self, 'networks', 'security-group', 'get')
+    cmd.flags['id'] = self.id
+    stdout, stderr, _ = cmd.Issue()
+    return not stderr
+
+
+class RackspaceSecurityGroupRule(resource.BaseResource):
+  """An object representing a Security Group Rule."""
+
+  def __init__(self, sec_group_rule_name, sec_group_id, direction=None,
+               ip_ver=None, protocol=None, port_range_min=None,
+               port_range_max=None, source_cidr=None):
+    super(RackspaceSecurityGroupRule, self).__init__()
+    self.id = None
+    self.name = sec_group_rule_name
+    self.sec_group_id = sec_group_id
+    self.direction = direction if direction in SEC_GROUP_DIRECTIONS else INGRESS
```

What you said above - "I already default to valid values here using the conditional assignments if someone passes a None value or something that is not valid" - is actually what I think the problem is. I think you should default to a valid value if someone passes None, but if someone passes something else that's invalid, you should raise an exception.

However, this is all internal, so it's up to you.